Speed up CI builds with ccache (seeded from develop)

marcocastignoli opened the pull request:
- This PR enables ccache for the main CMake build in CircleCI and stores the cache between runs.
- CircleCI now restores ~/.ccache on most branches from the latest cache produced on develop, so builds can reuse compiled objects.
- Runs on develop start from an empty cache and then save a fresh ~/.ccache snapshot (one per revision) to keep the seed cache updated.
- Adds/ensures ccache is installed on all relevant executors: macOS, Arch Linux, and the Ubuntu/emscripten buildpack-deps Docker images.
- The build step also resets and prints ccache stats (ccache -z / ccache -s) so we can see cache hits in the logs.

For the purpose of testing instead of creating the cache from `develop`, the cache is created by this branch `develop-cache-test`. We can later create another PR starting from this branch to see if the builds are indeed using the cache generated by `develop-cache-test`, before merging this PR let's remember to fix the occurrences of `develop-cache-test` in the code.

marcocastignoli committed:
Speed up CI builds with ccache

cameel commented on scripts/docker/buildpack-deps/Dockerfile.ubuntu2404:
```diff
@@ -39,6 +39,7 @@ RUN set -ex; \
 	apt-get install -qqy --no-install-recommends \
 		build-essential \
 		cmake \
+		ccache \
```
We have a special workflow for updating buildpack images and this needs to be done in two steps (separate PRs). See [`buildpack-deps/README.md`](https://github.com/argotorg/solidity/blob/develop/scripts/docker/buildpack-deps/README.md).

cameel commented on .circleci/config.yml:
```diff
@@ -208,9 +208,44 @@ commands:
 
   run_build:
     steps:
+      # TODO: remove the line below
+      # Before merging this PR we need to use an alternative develop branch to test the functionality, `develop-cache-test`
```
We want to be able to run without ccache on tags, so we still need to keep a non-cached variant of this command around.

marcocastignoli replied:
What do you think is the best solution to handle this case? Something like this would work or do we need a more generic solution?

```yml
      - when:
          condition:
            and:
              - equal: ["", << pipeline.git.tag >>]
          steps:
            - run:
                name: Build (ccache)
                command: |
                  export CCACHE_DIR="$HOME/.ccache"
                  export CCACHE_BASEDIR="$(pwd)"
                  export CCACHE_NOHASHDIR=1
                  mkdir -p "$CCACHE_DIR"
                  export CMAKE_OPTIONS="${CMAKE_OPTIONS:-} -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache"
                  ccache -z

                  scripts/ci/build.sh

                  ccache -s
      - when:
          condition:
            or:
              - not:
                  equal: ["", << pipeline.git.tag >>]
          steps:
            - run:
                name: Build (no ccache)
                command: scripts/ci/build.sh
```

cameel replied:
Yeah, that should work.

I mean, I would have preferred something that directly passes this decision as a parameter from the top, but I that would be annoying, because you'd have to modify most jobs to have parameters. A global version like this does not have this problem.

cameel commented on .circleci/config.yml:
```diff
@@ -208,9 +208,44 @@ commands:
+            export CCACHE_NOHASHDIR=1
+            mkdir -p "$CCACHE_DIR"
+            export CMAKE_OPTIONS="${CMAKE_OPTIONS:-} -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache"
+            ccache -z || true
```
Do we expect not to have this on any of the platforms?
Is that platform Windows? :P

cameel commented on .circleci/config.yml:
```diff
@@ -208,9 +208,44 @@ commands:
+                keys:
+                  # Reuse cache produced on develop-cache-test (prefix match; restores the most recent).
+                  - v1-ccache-{{ arch }}-develop-cache-test-
+                  - v1-ccache-{{ arch }}-
```
Why do we need `v1-ccache-{{ arch }}-` here? You only ever save the cache for `develop-cache-test` so the second line will never match.

Did you mean something like this:

```suggestion
            - restore_cache:
                keys:
                  # Reuse the cache for current branch or fall back to the one from develop-cache-test if not available.
                  # NOTE: prefix match; restores the most recent
                  - v1-ccache-{{ arch }}-{{ .Branch }}-
                  - v1-ccache-{{ arch }}-develop-cache-test-
```

along with adjusting the save part to this?
```yaml
      - save_cache:
          key: v1-ccache-{{ arch }}-{{ .Branch }}-{{ .Revision }}
```

This would always start a new branch with cache from `develop`, but then keep using and updating the cache for the branch.

This would likely let us reuse more, though not sure I like it, because it does mean that the cache for branches is never recreated from scratch (unless we just nuke the whole cache by bumping the version) and is even reused for a completely new branch that just happens to share the same name as an old one.

marcocastignoli replied:
Yeah, you are right it doesn't make sense to have `v1-ccache-{{ arch }}-` as fallback

cameel commented on .circleci/config.yml:
```diff
@@ -208,9 +208,44 @@ commands:
+            equal: ["develop-cache-test", << pipeline.git.branch >>]
+          steps:
+            - save_cache:
+                key: v1-ccache-{{ arch }}-develop-cache-test-{{ .Revision }}
```
I expect that we might want to invalidate cache whenever we suspect something weird is going on. It would be convenient to be able to do that without having to merge a PR. In an extreme case we might even not be able to merge one due to required jobs failing because of cache. In addition to the hard-coded version we could also keep a variable in CircleCI config that we can bump to get the same effect:

```suggestion
                key: v1-{{ .Environment.CCACHE_KEY }}-ccache-{{ arch }}-develop-cache-test-{{ .Revision }}
```

cameel commented on .circleci/config.yml:
```diff
@@ -208,9 +208,44 @@ commands:
+                  # Reuse cache produced on develop-cache-test (prefix match; restores the most recent).
+                  - v1-ccache-{{ arch }}-develop-cache-test-
+                  - v1-ccache-{{ arch }}-
       - run:
```
we usually add a warning here because not everyone may realize that changing the code between `restore_cache` and `save_cache` doe not invalidate the cache automatically.
```suggestion
                  - v1-ccache-{{ arch }}-
      # WARNING! If you edit anything between here and save_cache, remember to bump version or invalidate the cache manually.
      - run:
```

cameel commented on .circleci/config.yml:
```diff
@@ -208,9 +208,44 @@ commands:
+      - when:
+          condition:
+            not:
+              equal: ["develop-cache-test", << pipeline.git.branch >>]
```
The `breaking` branch needs special treatment as well. It's not a feature branch but currently would be treated like one. It's like `develop`, but for the next breaking version.

marcocastignoli committed:
Address PR comments
- improve comments and warnings
- treat ccache as always existing
- add CCACHE_KEY to easily invalidate cache
- skip cache also on "breaking" branch

marcocastignoli committed:
Bumo versions of buildpack-deps dockerfiles

marcocastignoli committed:
Skip ccache build when pushing tags

marcocastignoli committed:
Fix ccache permission error while testing buildpack-deps

marcocastignoli committed:
Update docker images to use version with ccache

marcocastignoli committed:
Fix cache key in restore_cache
